boa/src/object/object_.rs

```diff
@@ -0,0 +1,714 @@
+use super::GcObject;
+use crate::{
+    property::{AccessorDescriptor, Attribute, DataDescriptor, PropertyDescriptor, PropertyKey},
+    symbol::RcSymbol,
+    value::{RcString, Value},
+    BoaProfiler, Context, Result,
+};
+use rustc_hash::FxHashMap;
+
+use delegate::delegate;
+use enum_dispatch::enum_dispatch;
+
+#[enum_dispatch]
+pub(super) enum Object_ {
+    OrdinaryObject,
+    ExoticObject,
+}
+
+pub(super) struct OrdinaryObject {
+    indexed_properties: FxHashMap<u32, PropertyDescriptor>,
+    /// Properties
+    string_properties: FxHashMap<RcString, PropertyDescriptor>,
+    /// Symbol Properties
+    symbol_properties: FxHashMap<RcSymbol, PropertyDescriptor>,
+    /// Instance prototype `__proto__`.
+    prototype: Value,
+    /// Whether it can have new properties added to it.
+    extensible: bool,
+}
+
+impl CommonObject for OrdinaryObject {
+    fn proto(&self) -> Value {
+        self.prototype.clone()
+    }
+
+    fn ext(&self) -> bool {
+        self.extensible
+    }
+
+    fn set_proto(&mut self, proto: Value) {
+        self.prototype = proto
+    }
+
+    fn set_ext(&mut self, is_extensible: bool) {
+        self.extensible = is_extensible
+    }
+
+    fn get_prop<K>(&self, key: K) -> Option<&PropertyDescriptor>
+    where
+        K: Into<PropertyKey>,
```

Comment:
I'm still not sure if adding this to every function is the way to go. It improves the ergonomics of the api, yes, but now we can't take references when needed, like in this case. Maybe adding it only to functions that insert or create would be the ideal solution.